### PR TITLE
Create tabs manager and refactor tab related logic

### PR DESCRIPTION
**Pull request by victor-meng**
- created tabs manager class
- moved all tabs related logic into tabs manager
- all tab operation such as create, open, and delete are done in tabs manager

### CODE REVIEW

File: src/explorer.html [hunk 1]
```diff
@@ -187,171 +187,17 @@
...
-              </div>
-              <!-- /ko -->
-            </div>
-            <!-- Tabs Panes - End -->
-          </div>
-          <!-- Tabs and Tabs Panes - End -->
+          <tabs-manager
+            class="tabsContainer"
+            params="{data: tabsManager}"
+            data-bind="visible: tabsManager.openedTabs().length > 0"
```
Comment (languy) [inline src/explorer.html:199]: Shouldn't we always show the tab manager? If there's no tab, nothing should show up anyway, but conceptually, if, for example one day we have a [+] button to create a new tab in the tab manager, it should always show up.

File: src/Explorer/Explorer.ts [hunk 2]
```diff
@@ -1282,7 +1273,7 @@ export default class Explorer implements ViewModels.Explorer {
       class: "connectDialogButtons okBtn connectOkBtns",
       click: () => {
         $("#contextSwitchPrompt").dialog("close");
-        this.openedTabs([]); // clear all tabs so we dont leave any tabs from previous session open
+        this.tabsManager.clearAllTabs(); // clear all tabs so we dont leave any tabs from previous session open
```
[the remark this answers is not in the document]
Comment (victor-meng) [reply src/Explorer/Explorer.ts]: Done

File: src/Explorer/Explorer.ts [hunk 3]
```diff
@@ -2621,46 +2596,51 @@ export default class Explorer implements ViewModels.Explorer {
...
 
-    if (openedNotebookTabs.length > 0) {
-      openedNotebookTabs[0].onTabClick();
-      openedNotebookTabs[0].onActivate();
-      return true;
-    }
+    const notebookTabs: NotebookV2Tab[] = this.tabsManager.getTabs(
+      ViewModels.CollectionTabKind.NotebookV2,
+      (tab: ViewModels.Tab) => {
+        return (
```
[the remark this answers is not in the document]
Comment (victor-meng) [reply src/Explorer/Explorer.ts]: Done

File: src/Explorer/Explorer.ts [hunk 4]
```diff
@@ -2696,17 +2677,17 @@ export default class Explorer implements ViewModels.Explorer {
...
-          )
-          .forEach(tab => {
-            tab.tabTitle(newNotebookFile.name);
-            tab.tabPath(newNotebookFile.path);
-            (tab as NotebookV2Tab).notebookPath(newNotebookFile.path);
-          });
+        const notebookTabs: ViewModels.Tab[] = this.tabsManager.getTabs(
+          ViewModels.CollectionTabKind.NotebookV2,
+          (tab: NotebookV2Tab) => {
+            return tab.notebookPath && FileSystemUtil.isPathEqual(tab.notebookPath(), originalPath);
```
[the remark this answers is not in the document]
Comment (victor-meng) [reply src/Explorer/Explorer.ts]: Done

File: src/explorer.html (hunk 1 above)
Comment (victor-meng) [reply src/explorer.html]: That's a good point. I've removed the data-bind so the tabs manager is always visible.

File: src/Explorer/Tabs/TabsManager.ts [hunk 5]
```diff
@@ -0,0 +1,111 @@
...
+  public getTabs(
+    tabKind: ViewModels.CollectionTabKind,
+    comparator?: (tab: ViewModels.Tab) => boolean
+  ): ViewModels.Tab[] {
+    return this.openedTabs().filter((openedTab: ViewModels.Tab) => {
+      return openedTab.tabKind === tabKind && (!comparator || comparator(openedTab));
+    });
+  }
+
+  public closeNotebookTab(filepath: string): void {
```
[the remark this answers is not in the document]
Comment (victor-meng) [reply src/Explorer/Tabs/TabsManager.ts]: Done

**Pull request merged**